### PR TITLE
podman: determine cgroup paths automatically

### PR DESCRIPTION
**Pull request by bduffany**
* Move all the cgroup stuff from `podman.go` to a new package `cgroup.go` since it's growing a bit, and we also want to use this for Docker.
* Instead of requiring the cgroup flags, walk the /sys/fs/cgroup tree, searching for the cid in the path, and determine mem/cpu paths from that.
  * The dir scanning approach is used because there are several different variables that determine what the paths will actually be - whether systemd is used or not, whether we're on cgroup v1 or v2, whether podman is running rootless or not... maybe other stuff (that may possibly change in the future?) I figured it was too complicated and error-prone to try and take all of these things into account.
  * In terms of performance, it takes 8ms to scan the cgroupfs tree (tested on a GCP VM). It succeeds very quickly and manages to init the cgroup paths on the first try, even when running a NOP task (/bin/true). We only do the scan once, so this overhead is incurred only on the first task executed.

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/remote_execution/cgroup/cgroup.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+	memUsagePath := strings.ReplaceAll(p.MemoryTemplate, cidPlaceholder, cid)
+	cpuUsagePath := strings.ReplaceAll(p.CPUTemplate, cidPlaceholder, cid)
+
+	memUsageBytes, err := readInt64FromFile(memUsagePath)
+	if err != nil {
+		return nil, err
+	}
+	var cpuNanos int64
+	if p.CgroupVersion() == 1 {
```
Comment (iain-macdonald) [inline enterprise/server/remote_execution/cgroup/cgroup.go:61]: guard against 0 here?

File: enterprise/server/remote_execution/containers/podman/podman.go [hunk 2]
```diff
@@ -394,12 +400,19 @@ type podmanCommandContainer struct {
 	removed bool
 }
 
+// TODO: remove this and always use provider.New() instead.
 func NewPodmanCommandContainer(env environment.Env, image, buildRoot string, options *PodmanOptions) container.CommandContainer {
+	imageExistsCache, _ := newImageExistsCache()
```
Comment (iain-macdonald) [inline enterprise/server/remote_execution/containers/podman/podman.go:405]: I think you wanna pass this in from the Provider so it's shared across CommandContainers, right?

**Review by iain-macdonald — COMMENTED** [review]
couple nits but overall looks good!

File: enterprise/server/remote_execution/cgroup/cgroup.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/cgroup/cgroup.go:61]: find() returns an error in that case, so I think it's not needed.

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/podman/podman.go:405]: provider.New() doesn't call this func unfortunately - this func is only used for testing. Line 403 has a TODO to remove this func and only use the provider, even in tests

Comment (iain-macdonald) [reply enterprise/server/remote_execution/containers/podman/podman.go:409]: Oof, got me! Maybe one of us should get on that TODO soon 😛

File: enterprise/server/remote_execution/cgroup/cgroup.go (hunk 1 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/cgroup/cgroup.go:61]: lol i agree with @iain-macdonald 

i have a small brain and seeing:

```
if 1 {
} else if 2 {
} else {
 raise error(get out)
}
```

makes me feel safe

File: enterprise/server/remote_execution/cgroup/cgroup.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+func readInt64FromFile(path string) (int64, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return 0, err
+	}
```
Comment (tylerwilliams) [inline enterprise/server/remote_execution/cgroup/cgroup.go:161]: can you use `data, err := os.ReadFile(path)` here?

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 2 above)
Comment (tylerwilliams) [reply enterprise/server/remote_execution/containers/podman/podman.go:409]: +1

**Review by tylerwilliams — APPROVED** [review]
couple nits but looks great!

File: enterprise/server/remote_execution/containers/podman/podman.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/remote_execution/containers/podman/podman.go:409]: I can do it :) had started on it in this PR, but then it started getting super big so I split it into a separate branch

File: enterprise/server/remote_execution/cgroup/cgroup.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/remote_execution/cgroup/cgroup.go]: yep done

File: enterprise/server/remote_execution/cgroup/cgroup.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/remote_execution/cgroup/cgroup.go]: okay, fair :) done

**Pull request merged**